java/src/edu/ohsu/cslu/parser/chart/CellChart.java

```diff
@@ -182,38 +237,41 @@ public void updateInside(final Production p, final ChartCell leftCell, final Cha
                 final float insideProb) {
             final int nt = p.parent;
             if (viterbiMax && insideProb > getInside(nt)) {
-                if (bestEdge[nt] == null) {
-                    bestEdge[nt] = new ChartEdge(p, leftCell, rightCell);
-                } else {
-                    bestEdge[nt].prod = p;
-                    bestEdge[nt].leftCell = leftCell;
-                    bestEdge[nt].rightCell = rightCell;
-                }
+//                if (bestEdge[nt] == null) {
+//                    bestEdge[nt] = new ChartEdge(p, leftCell, rightCell);
+//                } else {
+//                    bestEdge[nt].prod = p;
+//                    bestEdge[nt].leftCell = leftCell;
+//                    bestEdge[nt].rightCell = rightCell;
+//                }
+                throw new NotImplementedException();
             }
             updateInside(nt, insideProb);
             updateCounts(p);
         }
 
         @Override
         public ChartEdge getBestEdge(final int nt) {
-            return bestEdge[nt];
+            return bestEdge.get(nt);
         }
 
         public List<ChartEdge> getBestEdgeList() {
             final List<ChartEdge> bestEdges = new LinkedList<ChartEdge>();
-            for (int i = 0; i < bestEdge.length; i++) {
-                if (bestEdge[i] != null) {
-                    bestEdges.add(bestEdge[i]);
+            for (int i = 0; i < bestEdge.length(); i++) {
+                ChartEdge e = bestEdge.get(i);
+                if (e != null) {
+                    bestEdges.add(e);
                 }
             }
             return bestEdges;
         }
 
         public boolean hasNT(final int nt) {
-            return inside[nt] > Float.NEGATIVE_INFINITY;
+            return bestEdge.get(nt) != null;
+//            return inside[nt] > Float.NEGATIVE_INFINITY;
         }
 
-        protected void addToHashSets(final int ntIndex) {
+        protected synchronized void addToHashSets(final int ntIndex) {
```

Comment:
should probably investigate removing the synchronized from this since it is likely one of the things that is limiting parallelism atm